Add TLS and certificate-based access security features

george-mcintyre opened the pull request:
### Add TLS and certificate-based access security features

Extended access security to support new fields (`METHOD`, `AUTHORITY`) and options (`ISTLS`) for certificate-based connection authorization. Introduced a new `RPC` access type and updated related APIs (`asAddClientX`, `asTrapWriteWithDataX`) for enhanced control. Adjusted tests and documentation to accommodate new functionalities and ensure backward compatibility.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
UNDO: Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add GitHub Actions workflow for multi-platform Docker builds

This workflow automates building and pushing EPICS base Docker images for multiple platforms (amd64, arm64). It uses repositories and configurations for both EPICS base and PVXS, triggered by manual dispatch or workflow calls.

george-mcintyre committed:
Add TLS and certificate-based access security features

https://github.com/george-mcintyre/pvxs/issues/74

Extended access security to support new fields (`METHOD`, `AUTHORITY`) and options (`ISTLS`) for certificate-based connection authorization. Introduced a new `RPC` access type and updated related APIs (`asAddClientX`, `asTrapWriteWithDataX`) for enhanced control. Adjusted tests and documentation to accommodate new functionalities and ensure backward compatibility.

george-mcintyre commented:
Replaces #653

mdavidsaver commented on modules/database/src/ioc/rsrv/camessage.c:
```diff
@@ -791,9 +791,11 @@ static int write_action ( caHdrLargeArray *mp,
     }
 
-    asWritePvt = asTrapWriteWithData ( pciu->asClientPVT,
+    asWritePvt = asTrapWriteWithDataX ( pciu->asClientPVT,
```
CA does not, and never will, have a concept of RPC call.  and imo. never will support TLS transport.  So I think the changes under rsrv/ should be removed as proof of API backwards compatibility.

george-mcintyre replied:
Agreed: Addressed in commit 56fdf388

mdavidsaver commented on modules/libcom/src/as/asLibRoutines.c:
```diff
@@ -63,22 +63,27 @@ static ASG *asAsgAdd(const char *asgName);
+ *    * the old structures are freed.
+ *
+ * @param arg Argument (not used)
+ */
```
Public API documentation comments should appear in the header, otherwise they will be ignored by doxygen.

george-mcintyre replied:
Resolved by 24c78119

mdavidsaver commented on modules/libcom/src/as/asLibRoutines.c:
```diff
@@ -361,8 +366,52 @@ void epicsStdCall asPutMemberPvt(ASMEMBERPVT asMemberPvt,void *userPvt)
+ *          `S_asLib_badMember` if the member is not provided,
+ *          `S_asLib_noMemory` if there is not enough memory to allocate the client structure,
+ *          or the status from `asComputePvt` which will be 0 if the client is successfully added
+ */
```
Rather than an almost copy paste for `asAddClient()` and `asAddClientX()`, I would prefer to document the former as being a short-hand for `asAddClientX(pasClientPvt, asMemberPvt, asl, user, "ca", NULL, host, 0)`.

george-mcintyre replied:
Addressed in f6d4c08e

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -92,6 +99,13 @@ LIBCOM_API long epicsStdCall asAddClient(
+    int asl,const char *user,char *method,char *authority,char *host, int isTLS);
+/*client must provide permanent storage for user, host, method, and authority */
+LIBCOM_API long epicsStdCall asChangeClientX(
+    ASCLIENTPVT asClientPvt,int asl,const char *user,char *method,char *authority, char *host, int isTLS);
```
I'm concerned about adding additional arguments to already long argument lists.  Especially when the new arguments are not simply appended.  Is it time to switch to passing in the arguments as a structure?

george-mcintyre replied:
I agree that we should change the signature but I hope we can address the format of ACF files to make them forwards compatible (future proof/extensible).  e.g. allowing them to syntactically accept arbitrary entries for `RULES` section but having the parser discard entries that it does not implement.  This would allow us to introduce changes like this one without breaking legacy clients.    At that time I think it would be good to change the signature because in any case it would be a breaking change for legacy clients (so why not do it at the same time)

mdavidsaver commented on modules/libcom/test/aslibtest.c:
```diff
@@ -111,6 +166,220 @@ static void testHostNames(void)
+    size_t size = 0;
+
+    // DEFAULT
+    FILE *fp = open_memstream(&buf, &size);
```
As @anjohnson [mentioned previously](https://github.com/epics-base/epics-base/pull/563#discussion_r1846982443), `open_memstream` is not portable to all supported targets.

It is also unnecessary.  This test is allowed to write files to `$PWD`.

george-mcintyre replied:
Resolved with #ce377d73

mdavidsaver reviewed (changes requested):
Some specific comments, including some from #563.  I also intend to write up a more conceptual discussion of what an extended credential model can/should be.

george-mcintyre committed:
https://github.com/epics-base/epics-base/pull/618#discussion_r2002141050

@mdavidsaver

george-mcintyre committed:
Refactor AS library comments and function declarations

Move detailed function comments from implementation files to header files.

https://github.com/epics-base/epics-base/pull/618#discussion_r2002141708

@mdavidsaver

george-mcintyre committed:
Refactor testRulesDumpOutput

Removed memory file operations in lieu of mor portability.

Extracted duplicate logic from testRulesDumpOutput into a helper function, runRestDumpRules, for improved readability and maintainability. Added a readFile utility function to handle file content reading and simplify temporary file handling. These changes streamline the testing process and reduce code duplication.

https://github.com/epics-base/epics-base/pull/618#discussion_r2002148444

@mdavidsaver

george-mcintyre committed:
Refactor ASG client function documentation for clarity.

Simplified documentation of legacy API functions (`asAddClient` and `asChangeClient`) for "ca" behavior. Updated documentation to clarify usage and align with extended versions (`asAddClientX` and `asChangeClientX`).

https://github.com/epics-base/epics-base/pull/618#discussion_r2002144003

@mdavidsaver

george-mcintyre commented:
Replaced with #641
